pyccl/base.py

```diff
@@ -298,3 +299,345 @@ def increment(self):
+
+    Arguments:
+        func (``function``):
+            Function which changes one of its ``CCLObject`` arguments.
+        argv (``int``):
+            Which argument should be unlocked. Defaults to the first argument.
+            This should be a ``CCLObject``, or the decorator will do nothing.
+        mutate (``bool``):
+            If after the function ``instance_old != instance_new``, the
+            instance is mutated. If ``True``, the representation of the
+            object will be reset.
+    """
+    if func is None:
+        # called with parentheses
+        return functools.partial(unlock_instance, argv=argv, mutate=mutate)
+
+    @functools.wraps(func)
+    def wrapper(*args, **kwargs):
+        # Pick argument from list of `args` or `kwargs` as needed.
+        size = len(args)
+        arg = args[argv] if size > argv else list(kwargs.values())[argv-size]
+        with UnlockInstance(arg, mutate=mutate):
+            out = func(*args, **kwargs)
+        return out
+    return wrapper
+
+
+class CCLObjectMeta(ABCMeta):
+    """Metaclass for ``CCLObject``.
+
+    Enables linking of abstract methods, if ``__linked_abstractmethods__``
+    is provided for the base class. Subclasses that define either of
+    the linked methods will satisfy the abstraction requirement.
+
+    Example:
+        Subclasses of the following class can be instantiated if either
+        ``method1`` or ``method2`` (and ``another_method``) are defined.
+        Otherwise, it falls back to normal ``abc.ABCMeta`` behavior:
+
+        >>> class MyBaseClass(metaclass=ABCMeta):
+                __linked_abstractmethods__ = 'method1', 'method2'
+                @abstractmethod
+                def method1(self):
+                    ...
+                @abstractmethod
+                def method2(self):
+                    ...
+                @abstractmethod
+                def another_method(self):
+                    ...
+    """
+
+    def __new__(mcls, name, bases, clsdict):
+        """Class factory for ``CCLObject``.
+
+        .. note:: Naming convention: ``mcls`` -- metaclass, ``cls`` -- class
+        """
+        cls = super().__new__(mcls, name, bases, clsdict)
+
+        # Check if abstraction criteria are satisfied.
+        if hasattr(cls, "__linked_abstractmethods__"):
+            # Tap into class creation and remove all linked abstract methods
+            # from the `__abstractmethods__` hook.
+            linked = cls.__linked_abstractmethods__
+
+            def is_abstract(cls, method):
+                # Check if the `method` of class `cls` is abstract.
+                meth = getattr(cls, method)
+                return getattr(meth, "__isabstractmethod__", False)
+
+            if not all([is_abstract(cls, method) for method in linked]):
+                abstracts = set(cls.__abstractmethods__) - set(linked)
+                cls.__abstractmethods__ = frozenset(abstracts)
```

Comment:
Actually I agree. It was already in my to-do list of things to eliminate so for the past few days I'd been trying to think of ways to do this more elegantly.
The `abc` module introduced the really convenient `update_abstractmethods` which can be used as a class decorator and could achieve what I was trying to implement. The problem is this only exists in Python 3.10. 🙄

So I'll probably be dropping this altogether and just add a check after the class is defined, that verifies that `_real` or `_fourier` exist. Do you agree?

Reply:
I had a very close look at this and (unfortunately) metaclasses is the only obvious option if we want to link abstract methods. This can't be done any other way (e.g. within `__init_subclass__`) because `abc` only determines which methods are abstract after all this has run. So the only way to tap into the inner workings of `abc` is to use a metaclass (abstract methods are defined using `ABCMeta` in the first place).

Seeing as we only want this to link `_real` and `_fourier` I also find no point in having all this extra code for these two methods, so I removed them completely.